File: merge-intervals/merge_intervals.py
# Merge Intervals

# Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals, and 
# return an array of the non-overlapping intervals that cover all the intervals in the input.

# Example 1:
# 
# Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
# Output: [[1,6],[8,10],[15,18]]
# Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        _intervals= []
        for i in intervals:
            if not _intervals or _intervals[-1][1] < i[0]:
                _intervals.append(i)
            else:
                _intervals[-1][1] = max(_intervals[-1][1], i[1])
        
        print(_intervals)
        return _intervals

File: merge-intervals/merge_intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
	if (intervals.length == 0) {
		return []
	}
	if (intervals.length == 1) {
		return intervals
	}

	let ans = []
	intervals = intervals.sort(function(a, b) {
		return a[0] - b[0];
	});
	ans.push(intervals[0]);
	for (let i = 1; i <= intervals.length - 1; i++) {
		if (ans[ans.length - 1][1] >= intervals[i][0]) {
			if (ans[ans.length - 1][1] <= intervals[i][1]) {
                ans[ans.length - 1][1] = intervals[i][1]
			} 
		} else {
			ans.push(intervals[i])
		}
	}
	return ans;
};
